### PR TITLE
Adds recursive `__all__` dictionary

### PR DESCRIPTION
**Pull request by arjxn-py**
# Description

Been proposed in #3866 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by arjxn-py** [comment]
Also related to #2427

**Review by agriyakhetarpal — COMMENTED** [review]
Thanks, @arjxn-py – overall, I believe that adding `__all__` to the codebase makes a lot of sense, it's the more modern way of bringing structure to the public API, plus providing efficient tab completions for IDEs and text editors.

I wonder if doing this in a staggering manner, i.e., across multiple `pybamm/module/submodule/__init__.py` files makes more sense, in comparison to adding them in the same file? We have way too many changes in a single file right now, it would become difficult to keep track of where these symbols and attributes come from (some in-line comments could also be of help). Keeping everything split across files will be just a one-time change to review (though it would have to be thorough), and then adding to items to multiple `__all__` lists will be simpler, too, since they will be smaller. I suppose you're using the `mkinit` tool right now that has come up in our previous discussions – the main issue with it was that we would have needed to ask contributors to run a command every time they modified something for the API, and we don't want that additional complexity. Splitting this across multiple files should make the structure easier to follow and then manually modify on future changes, without the use of the tool.

File: pybamm/solvers/c_solvers/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+__all__ = []
```
Comment (agriyakhetarpal) [inline pybamm/solvers/c_solvers/__init__.py:1]: This one is mentioned by CodeCov as a line we have missing coverage for, so I think we can remove the line.

```suggestion
```

**Review by agriyakhetarpal — COMMENTED** [review]
Thanks, @arjxn-py! I think this looks all good to me, unless there is something more that you plan to do here and push further changes.

**Comment by arjxn-py** [comment]
Planned to add some directives to contribution guide as well but I believe it should be fine without that, we just have to little bit more careful on the PRs modifying method & attribute names or adding new of them.

**Comment by arjxn-py** [comment]
Feel free to merge it after approval.

**Review by agriyakhetarpal — APPROVED** [review]
Looks fine to me, @arjxn-py! I'll request another review from @Saransh-cpp, we can merge soon after.

**Comment by kratman** [comment]
Is this something that needs to be updated frequently?

**Comment by agriyakhetarpal** [comment]
No, as long as we're not changing the API frequently. It does need to be updated when adding new features, though.

The goal will be that once we have a stable public-private API demarcation, `__all__` will reflect the public API.

**Comment by arjxn-py** [comment]
Just a gentle ping to @Saransh-cpp to review this and go forward with it if looks fine, Thanks :)

**Review by Saransh-cpp — APPROVED** [review]
Thanks, @arjxn-py! Looks good to me!

**Pull request merged**